File: find_symbol.ts
/**
 * find_symbol — Locate where a symbol is defined in the workspace.
 *
 * Given a name (class, interface, function, method), scans the workspace AST
 * and returns all files + lines where that symbol is defined.
 *
 * Language-aware: only loads the tree-sitter deps needed for the selected language.
 * Supports exact match (default) and fuzzy match (contains).
 *
 * Supports: Java (Spring hexagonal), TypeScript/TSX/JavaScript/JSX (React Router 7).
 */

import { tool } from "@opencode-ai/plugin"
import * as path from "node:path"
import * as fs from "node:fs/promises"

const SCRIPT = path.join(
  path.dirname(new URL(import.meta.url).pathname),
  "find_symbol.py",
)

function uvDepsForLanguage(language: string): string[] {
  const base = ["--with", "tree-sitter"]
  switch (language) {
    case "java":
      return [...base, "--with", "tree-sitter-java"]
    case "ts":
    case "typescript":
      return [...base, "--with", "tree-sitter-typescript"]
    case "auto":
    default:
      return [...base, "--with", "tree-sitter-java", "--with", "tree-sitter-typescript"]
  }
}

export default tool({
  description: `Locate where a symbol is defined in the workspace using AST analysis (tree-sitter).

Given a name, scans the workspace and returns all files + lines where that class, interface,
function, method, or type is defined. Returns only file + line + kind — no file content.

Use this when you know a symbol name but not its file. Much cheaper than grep or manual search.

Supports exact match (default) and fuzzy/contains match.
Supports language filter to avoid loading unnecessary parsers.

Supports: Java (Spring hexagonal), TypeScript/TSX/JavaScript/JSX (React Router 7).

Examples:
  find_symbol(name="CreateTitularUseCase", language="java")
  find_symbol(name="getTitularById", language="java", kind="method")
  find_symbol(name="useSession", language="ts")
  find_symbol(name="Titular", language="java", fuzzy=true)
  find_symbol(name="loader", language="ts", kind="function")`,

  args: {
    name: tool.schema
      .string()
      .describe("Symbol name to search for. Exact match by default. Use fuzzy=true for partial/contains match."),

    language: tool.schema
      .enum(["auto", "java", "ts", "typescript"])
      .optional()
      .describe("Language to search in. 'auto' searches all supported files. Prefer specifying a language to speed up the scan. Defaults to 'auto'."),

    kind: tool.schema
      .enum(["any", "class", "interface", "function", "method", "type", "enum", "constructor", "annotation"])
      .optional()
      .describe("Filter results by definition kind. Defaults to 'any' (all kinds)."),

    fuzzy: tool.schema
      .boolean()
      .optional()
      .describe("If true, match names that CONTAIN the search term (case-insensitive). Default: false (exact match)."),
  },

  async execute(args, context) {
    const workspace = context.directory ?? process.cwd()
    const resolvedWorkspace = path.resolve(workspace)

    try {
      await fs.access(SCRIPT)
    } catch {
      return `find_symbol.py not found at: ${SCRIPT}`
    }

    const language = args.language ?? "auto"
    const kind = args.kind ?? "any"
    const fuzzy = args.fuzzy ?? false
    const uvDeps = uvDepsForLanguage(language)

    let stdout = ""
    let stderr = ""

    try {
      const proc = Bun.spawn(
        [
          "uv", "run",
          ...uvDeps,
          "python3", SCRIPT,
          resolvedWorkspace,
          args.name,
          JSON.stringify({ language, kind, fuzzy }),
        ],
        {
          stdout: "pipe",
          stderr: "pipe",
          cwd: workspace,
        },
      )

      stdout = await new Response(proc.stdout).text()
      stderr = await new Response(proc.stderr).text()
      const exitCode = await proc.exited

      if (exitCode !== 0) {
        return `find_symbol failed (exit ${exitCode})\n\n${stderr || stdout}`
      }
    } catch (e) {
      return `Failed to run uv: ${e instanceof Error ? e.message : String(e)}\n\nMake sure 'uv' is installed and in PATH.`
    }

    let result: {
      matches?: Array<{ file: string; line: number; kind: string; name: string }>
      count?: number
      error?: string
      trace?: string
      options?: { language?: string; kind?: string; fuzzy?: boolean }
    }

    try {
      result = JSON.parse(stdout.trim())
    } catch {
      return `Could not parse output:\n${stdout}\n\nStderr:\n${stderr}`
    }

    if (result.error) {
      return `Error:\n${result.error}${result.trace ? `\n\nTraceback:\n${result.trace}` : ""}`
    }

    if (!result.matches || result.matches.length === 0) {
      const hint = fuzzy
        ? `No definitions found containing '${args.name}'`
        : `No definitions found for '${args.name}'. Try fuzzy=true for partial matches.`
      return hint
    }

    const lines = result.matches.map(
      (m, i) => `  ${i + 1}. ${m.file}:${m.line}  [${m.kind}] ${m.name}`,
    )

    const filterParts: string[] = []
    if (language !== "auto") filterParts.push(`language: ${language}`)
    if (kind !== "any") filterParts.push(`kind: ${kind}`)
    if (fuzzy) filterParts.push("fuzzy: true")
    const filterStr = filterParts.length > 0 ? `\nFilters: ${filterParts.join(", ")}` : ""

    return [
      `Symbol: \`${args.name}\``,
      `Matches: ${result.count}${filterStr}`,
      "",
      lines.join("\n"),
      "",
      "Use read or scan_module to inspect the file content.",
    ].join("\n")
  },
})

File: find_symbol.py
#!/usr/bin/env python3
"""
find_symbol.py — Locate where a symbol is defined in the workspace.

Given a name (class, interface, function, method), scans the workspace AST
and returns all files + lines where that symbol is defined.

Accepts: <workspace> <name> [options_json]

Options JSON (optional):
  language: string (default "auto") — "auto", "java", "ts"/"typescript"
  kind:     string (default "any")  — filter by kind: "class", "interface",
                                      "function", "method", "type", "any"
  fuzzy:    bool   (default false)  — if true, match names containing <name>
                                      (case-insensitive) instead of exact match
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

_TOOLS_DIR = str(Path(__file__).resolve().parent)
if _TOOLS_DIR not in sys.path:
    sys.path.insert(0, _TOOLS_DIR)


# ---------------------------------------------------------------------------
# Node kind constants
# ---------------------------------------------------------------------------

JAVA_DEFINITION_NODES = {
    "class_declaration": "class",
    "interface_declaration": "interface",
    "enum_declaration": "enum",
    "annotation_type_declaration": "annotation",
    "method_declaration": "method",
    "interface_method_declaration": "method",
    "constructor_declaration": "constructor",
}

TS_DEFINITION_NODES = {
    "class_declaration": "class",
    "interface_declaration": "interface",
    "type_alias_declaration": "type",
    "function_declaration": "function",
    "method_definition": "method",
    "abstract_method_signature": "method",
    "enum_declaration": "enum",
}

# TS variable/const that declare functions or components
TS_VARIABLE_NODES = {
    "variable_declarator",
}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _name_matches(candidate: str, target: str, fuzzy: bool) -> bool:
    if fuzzy:
        return target.lower() in candidate.lower()
    return candidate == target


def _kind_matches(node_kind: str, filter_kind: str) -> bool:
    if filter_kind == "any":
        return True
    return node_kind == filter_kind


# ---------------------------------------------------------------------------
# Java searcher
# ---------------------------------------------------------------------------


def _search_java(
    path: Path,
    workspace: Path,
    name: str,
    kind_filter: str,
    fuzzy: bool,
) -> list[dict]:
    from trace_core.common import node_text, parse, walk

    tree, src = parse(path)
    if not tree:
        return []

    results: list[dict] = []

    for node_type, node_kind in JAVA_DEFINITION_NODES.items():
        if not _kind_matches(node_kind, kind_filter):
            continue
        for node in walk(tree.root_node, node_type):
            # Name is always an "identifier" child
            for child in node.children:
                if child.type == "identifier":
                    candidate = node_text(child, src)
                    if _name_matches(candidate, name, fuzzy):
                        results.append(
                            {
                                "file": str(path.relative_to(workspace)),
                                "line": node.start_point[0] + 1,
                                "kind": node_kind,
                                "name": candidate,
                            }
                        )
                    break

    return results


# ---------------------------------------------------------------------------
# TypeScript/JavaScript searcher
# ---------------------------------------------------------------------------


def _search_ts(
    path: Path,
    workspace: Path,
    name: str,
    kind_filter: str,
    fuzzy: bool,
) -> list[dict]:
    from trace_core.common import node_text, parse, walk

    tree, src = parse(path)
    if not tree:
        return []

    results: list[dict] = []

    # Named declaration nodes (class, interface, type, function, enum)
    for node_type, node_kind in TS_DEFINITION_NODES.items():
        if not _kind_matches(node_kind, kind_filter):
            continue
        for node in walk(tree.root_node, node_type):
            for child in node.children:
                if child.type == "identifier" or child.type == "type_identifier":
                    candidate = node_text(child, src)
                    if _name_matches(candidate, name, fuzzy):
                        results.append(
                            {
                                "file": str(path.relative_to(workspace)),
                                "line": node.start_point[0] + 1,
                                "kind": node_kind,
                                "name": candidate,
                            }
                        )
                    break

    # const/let/var declarations that export functions or components
    if _kind_matches("function", kind_filter) or kind_filter == "any":
        for decl_node in walk(
            tree.root_node, "lexical_declaration", "variable_declaration"
        ):
            for var_decl in walk(decl_node, "variable_declarator"):
                name_node = var_decl.child_by_field_name("name")
                if not name_node:
                    continue
                candidate = node_text(name_node, src)
                if not _name_matches(candidate, name, fuzzy):
                    continue
                # Only include if the value is a function/arrow/component
                value_node = var_decl.child_by_field_name("value")
                if value_node and value_node.type in (
                    "arrow_function",
                    "function_expression",
                    "function",
                    "jsx_element",
                ):
                    results.append(
                        {
                            "file": str(path.relative_to(workspace)),
                            "line": var_decl.start_point[0] + 1,
                            "kind": "function",
                            "name": candidate,
                        }
                    )

    return results


# ---------------------------------------------------------------------------
# Language initializer
# ---------------------------------------------------------------------------


def _init_languages(language: str):
    if language in ("auto", "java"):
        try:
            from trace_core.analyzers import java  # noqa: F401
        except ImportError:
            if language == "java":
                print(json.dumps({"error": "Java tree-sitter parser not available."}))
                sys.exit(1)

    if language in ("auto", "ts", "typescript"):
        try:
            from trace_core.analyzers import typescript  # noqa: F401
        except ImportError:
            if language in ("ts", "typescript"):
                print(
                    json.dumps(
                        {"error": "TypeScript tree-sitter parser not available."}
                    )
                )
                sys.exit(1)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------


def main():
    if len(sys.argv) < 3:
        print(
            json.dumps(
                {"error": "Usage: find_symbol.py <workspace> <name> [options_json]"}
            )
        )
        sys.exit(1)

    workspace_str = sys.argv[1]
    name = sys.argv[2]

    options: dict = {}
    if len(sys.argv) >= 4:
        try:
            options = json.loads(sys.argv[3])
        except json.JSONDecodeError:
            pass

    language = str(options.get("language", "auto")).lower()
    kind_filter = str(options.get("kind", "any")).lower()
    fuzzy = bool(options.get("fuzzy", False))

    _init_languages(language)

    from trace_core.common import SUPPORTED, should_skip

    workspace = Path(workspace_str).resolve()

    # Collect files to scan based on language filter
    ext_map: dict[str, set[str]] = {
        "java": {".java"},
        "ts": {".ts", ".tsx", ".js", ".jsx"},
        "typescript": {".ts", ".tsx", ".js", ".jsx"},
        "auto": set(SUPPORTED.keys()),
    }
    allowed_exts = ext_map.get(language, set(SUPPORTED.keys()))

    matches: list[dict] = []

    for p in sorted(workspace.rglob("*")):
        if not p.is_file():
            continue
        if p.suffix.lower() not in allowed_exts:
            continue
        if should_skip(p):
            continue

        if p.suffix.lower() == ".java":
            found = _search_java(p, workspace, name, kind_filter, fuzzy)
        else:
            found = _search_ts(p, workspace, name, kind_filter, fuzzy)

        matches.extend(found)

    print(
        json.dumps(
            {
                "matches": matches,
                "count": len(matches),
                "options": {
                    "language": language,
                    "kind": kind_filter,
                    "fuzzy": fuzzy,
                },
            }
        )
    )


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        import traceback

        print(json.dumps({"error": str(exc), "trace": traceback.format_exc()}))
        sys.exit(1)
